ui/components/TracePage/TracePage.tsx

```diff
@@ -277,6 +287,23 @@ const TreeProvider = ({ traceId, children }: { traceId: string; children: ReactN
     return undefined;
   }
 
+  const isVisible = useMemo(() => {
+    if (!hideOthers || !highlighted) return () => true;
+
+    const checkParents = memoize((id: string): boolean => {
+      const call = calls[id];
+      return call && (isHighlighted(call, highlighted) || checkParents(call.parent));
+    });
+    const checkChildren = memoize((id: string): boolean => {
+      const call = calls[id];
+      return (
+        call &&
+        (isHighlighted(call, highlighted) || Object.keys(call.children || {}).some(checkChildren))
+      );
+    });
+    return (id: string) => checkParents(id) || checkChildren(id);
+  }, [hideOthers, highlighted, calls]);
```

Comment:
Nice uses of short-circuiting. If we decide to keep this code, this should be extracted so it can be tested